src/Silicon.cpp

```diff
@@ -561,27 +489,95 @@ namespace galsim {
                 int polyi2 = std::min(i + (_qDist+1), i2);
                 int polyj1 = std::max(j - (_qDist+1), j1);
                 int polyj2 = std::min(j + (_qDist+1), j2);
-                int disti = nxCenter + polyi1 - i;
 
-                for (int polyi=polyi1; polyi<=polyi2; ++polyi, ++disti) {
-                    int distj = nyCenter + polyj1 - j;
-                    int index = (polyi - i1) * ny + (polyj1 - j1);
-                    int dist_index = disti * _ny + distj;
+                // work out indices of update region in horizontal linear array
+                int horizBoundaryX = (((i - i1) - _qDist) * horizontalPixelStride());
+                int horizDistortionX = ((nxCenter - _qDist) * horizontalPixelStride());
+                int horizW = (((_qDist * 2) + 1) * horizontalPixelStride());
+                if (horizBoundaryX < 0) {
+                    // handle case where stencil goes off left side
+                    horizW += horizBoundaryX;
+                    horizDistortionX -= horizBoundaryX;
+                    horizBoundaryX = 0;
+                }
+                if ((horizBoundaryX + horizW) > horizontalRowStride(nx)) {
+                    // handle case where stencil goes off right side
+                    horizW = horizontalRowStride(nx) - horizBoundaryX;
+                }
 
-                    for (int polyj=polyj1; polyj<=polyj2; ++polyj, ++distj, ++index, ++dist_index) {
-#if 0
-                        Polygon& distortion = _distortions[dist_index];
-                        Polygon& imagepoly = _imagepolys[index];
+                int horizBoundaryY = (j - j1) - _qDist;
+                int horizDistortionY = nyCenter - _qDist;
+                int horizH = (_qDist * 2) + 2;
+                if (horizBoundaryY < 0) {
+                    horizH += horizBoundaryY;
+                    horizDistortionY -= horizBoundaryY;
+                    horizBoundaryY = 0;
+                }
+                if ((horizBoundaryY + horizH) > ny) {
+                    horizH = ny - horizBoundaryY;
+                }
+
+                // update horizontal boundaries
+                for (int y = 0; y < horizH; ++y) {
+                    int index = ((horizBoundaryY + y) * horizontalRowStride(nx)) + horizBoundaryX;
+                    int dist_index = ((horizDistortionY + y) * horizontalRowStride(_nx)) + horizDistortionX;
+                    for (int x = 0; x < horizW; ++x, ++index, ++dist_index) {
+#ifdef _OPENMP
+#pragma omp atomic
 #endif
+                        _horizontalBoundaryPoints[index].x += _horizontalDistortions[dist_index].x * charge;
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
```

Comment:
I would think these atomic pragmas are slowing things down in a parallel context.  I think we could flip the order of loops here to have the outer loop be over the boundary points arrays, which would be the ones we would parallelize over.  Then each thread would have a different array, so they couldn't clobber each other on writes. The inner loops would be over the pixels, which could access the same pixels as each other sometimes, but only in a read capacity, so no race conditions.  That strikes me as likely to be faster.

Reply:
Ah, I see what you mean. Yes, that might well be faster, certainly for the cases where most pixels have some level of charge in them. The cost of the atomics is indeed quite high (I timed running the tests with the atomic pragmas disabled and they ran 2-3x faster, obviously with some errors due to the race condition, but gives a good idea of how much they slow things down). I'll code up a version with the loops flipped and see how fast it goes.